[Form Control Refresh][visionOS] Improve gaze glow for updated native control shapes

lilyspiniolas opened the pull request:
#### ca289cd2b10c7f52a3f71791ada7d1a413148dea
<pre>
[Form Control Refresh][visionOS] Improve gaze glow for updated native control shapes
<a href="https://bugs.webkit.org/show_bug.cgi?id=295388">https://bugs.webkit.org/show_bug.cgi?id=295388</a>
<a href="https://rdar.apple.com/154134813">rdar://154134813</a>

Reviewed by Aditya Keerthi.

In RenderThemeCocoa, controls which should have adjusted interaction
regions now each have a method which returns the shape of the control.
These methods are used both during painting, and during the computation
of interaction regions.

In `interactionRegionForRenderedRegion`, we now check if the region
belongs to a control with native appearance. If it does, we
call new method `RenderThemeCocoa::shapeForInteractionRegion`, which
provides information about the control&apos;s shape that we can use in order
to adjust the interaction region. This includes the size, corner radius,
path (if needed, only required if a clip path is present), and whether or
not the region should be rendered using continuous corners or typical
rounded corners. When a clip path is already present, we update it to be
the intersecting path between the existing clip path and the bounding path
for the control.

Added layout tests LayoutTests/interaction-region/form-control-refresh/*
which test the adjusted region shapes for native controls by dumping the
regions for a native control, a zoomed native control, a clipped native
control, and a native control that is both zoomed and clipped.

* LayoutTests/interaction-region/button-in-link-expected.txt:
* LayoutTests/interaction-region/clip-path-with-label-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/form-control-refresh/button-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/button-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/checkbox-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/checkbox-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/date-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/date-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/radio-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/radio-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/search-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/search-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/select-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/select-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/slider-thumb-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/slider-thumb-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/switch-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/switch-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/textfield-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/textfield-native-interaction-region.html: Added.
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/labels-expected.txt:
* LayoutTests/interaction-region/select-element-expected.txt:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/platform/visionos-2/TestExpectations:
* LayoutTests/platform/visionos-2/interaction-region/button-in-link-expected.txt: Copied from LayoutTests/interaction-region/button-in-link-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/clip-path-with-label-expected.txt: Copied from LayoutTests/interaction-region/clip-path-with-label-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/event-region-overflow-expected.txt: Copied from LayoutTests/interaction-region/event-region-overflow-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/icon-inside-button-single-region-expected.txt: Copied from LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/input-type-file-region-expected.txt: Copied from LayoutTests/interaction-region/input-type-file-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/input-type-range-region-expected.txt: Copied from LayoutTests/interaction-region/input-type-range-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/labels-expected.txt: Copied from LayoutTests/interaction-region/labels-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/select-element-expected.txt: Copied from LayoutTests/interaction-region/select-element-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/text-input-expected.txt: Copied from LayoutTests/interaction-region/text-input-expected.txt.
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::drawRect):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::addContinuousRoundedRect):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustedPaintRect const):
(WebCore::paintCheckboxRadioInnerShadowForVectorBasedControls):
(WebCore::continuousRoundedShape):
(WebCore::roundedShape):
(WebCore::nodeIsDateOrTimeRelatedInput):
(WebCore::shapeForButton):
(WebCore::shapeForCheckbox):
(WebCore::shapeForRadioButton):
(WebCore::searchFieldCanBeCapsule):
(WebCore::shapeForSearchField):
(WebCore::shapeForSwitchTrack):
(WebCore::shapeForSliderThumb):
(WebCore::shapeForTextAreaOrTextField):
(WebCore::RenderThemeCocoa::paintCheckboxForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintRadioForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSliderThumbForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSearchFieldForVectorBasedControls):
(WebCore::RenderThemeCocoa::shapeForInteractionRegion):
(WebCore::RenderThemeCocoa::inflateRectForInteractionRegion):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustedPaintRect const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):

Canonical link: <a href="https://commits.webkit.org/297559@main">https://commits.webkit.org/297559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90497fc78cd9f252e12028260e5e6a2ece8d5842

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112123 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118150 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62381 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114085 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40375 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85187 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35852 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25942 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100890 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65616 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25249 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19029 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62005 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95329 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19104 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121480 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39161 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29156 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94017 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39542 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97132 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93840 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23988 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39053 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16840 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35227 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44573 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38689 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42020 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40405 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -632,8 +638,40 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
+
+            maskedCorners = { };
+            cornerRadius = 0;
+            clipPath = adjustedPath;
```
`clipPath` is kind of a last resort solution for gaze glow (expensive on many levels).
I'm pretty worried we could end up with most form controls requiring clipPath (and then CAShapeLayer masking on the UI side).

Can we add a `useContinuousCornerCurve` boolean on the InteractionRegion struct instead?
(and then set it up properly [here](https://github.com/WebKit/WebKit/blob/12938589f723f90a88f566ff6ba314aa4782cd79/Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm#L302))

Looks like it could cover a lot of cases and greatly help!

Even if it breaks
```
// If a clip path is already present, find the intersecting path for the corrected path and the
// current clip path. The result will be the new clip path.
```
which is neat, I think we'd still want to do it.

etiennesegonzac commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -632,14 +638,63 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
+        const auto needsPath = clipPath ? ShouldComputePath::Yes : ShouldComputePath::No;
+        if (auto themeShape = RenderThemeCocoa::shapeForInteractionRegion(*regionRendererBox, regionRendererBox->borderBoxRect(), needsPath)) {
+            if (themeShape->cornerType == CornerType::Continuous)
+                useContinuousCorners = true;
```
should we bail in `EventRegionContext::shrinkWrapInteractionRegions()` when a region uses continuous corners?

lilyspiniolas replied:
Bailing early seems to result in gaze glow no longer appearing for labels associated with the controls. To preserve the updated shape in the case of associated labels, I added a check in `interactionRegionForRenderedRegion` which sets a clip path if the element has an associated label.

etiennesegonzac replied:
Interesting!
Wonder if the `hasLabel` is the right check though, or if it's a general issue with multiple regions sharing a `nodeIdentifier` (which we can only know once we're at the shrink wrap phase).

The safest thing might be to detect the continuous corners in `shrinkWrapInteractionRegions` and bail out _by keeping each region separated_. The control will get the continuous rect, and the label it's own classic region.

lilyspiniolas replied:
I see, I added some updated logic which avoids shrink wrapping regions with continuous corners and instead adds their paths to the discovered clip paths. Is that sufficient?

etiennesegonzac replied:
yes that'll work!

etiennesegonzac reviewed (commented):
The InteractionRegion part (and the tests!) look good!
The RenderTheme part probably needs more eyes from reviewers :)

aprotyas commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -603,8 +609,8 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         clipPath = path;
-    } else if (const auto& renderBox = dynamicDowncast<RenderBox>(regionRenderer)) {
-        auto borderShape = BorderShape::shapeForBorderRect(renderBox->style(), renderBox->borderBoxRect());
+    } else if ((regionRendererBox = dynamicDowncast<RenderBox>(regionRenderer))) {
```
You don't need the extra parentheses right?
```suggestion
    } else if (regionRendererBox = dynamicDowncast<RenderBox>(regionRenderer)) {
```

lilyspiniolas replied:
Without them there's a build error "using the result of an assignment as a condition without parentheses"

aprotyas commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -670,9 +728,14 @@ TextStream& operator<<(TextStream& ts, const InteractionRegion& interactionRegio
+#if ENABLE(FORM_CONTROL_REFRESH)
+            auto useContinuousCorners = interactionRegion.useContinuousCorners;
+            if (useContinuousCorners)
+                ts.dumpProperty("useContinuousCorners"_s, useContinuousCorners);
```
maybe? Since you're only ever logging the true case


```suggestion
            if (interactionRegion.useContinuousCorners)
                ts.dumpProperty("useContinuousCorners"_s, true);
```

aprotyas commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -200,6 +200,17 @@ void Path::addRoundedRect(const LayoutRoundedRect& rect)
+    if (isEmpty())
+        m_data = PathSegment(PathContinuousRoundedRect { rect, cornerRadius, cornerRadius });
+    else
+        ensureProtectedImpl()->add(PathContinuousRoundedRect { rect, cornerRadius, cornerRadius });
```
maybe, since you're creating the rounded rect instance anyway?
```suggestion
    PathContinuousRoundedRect roundedRect { rect, cornerRadius, cornerRadius };
    if (isEmpty())
        m_data = PathSegment(WTFMove(roundedRect));
    else
        ensureProtectedImpl()->add(WTFMove(roundedRect));
```

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
```diff
@@ -35,10 +35,36 @@ struct AttachmentLayout;
 
+#if ENABLE(FORM_CONTROL_REFRESH)
+
+enum class CornerType {
```
We can have tighter packing
```suggestion
enum class CornerType : uint8_t {
```

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
```diff
@@ -35,10 +35,36 @@ struct AttachmentLayout;
+    CornerType cornerType = CornerType::Noncontinuous;
+};
+
+enum class ShouldComputePath {
```
```suggestion
enum class ShouldComputePath : bool {
```

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+
+    Path path;
+    path.addRoundedRect(FloatRoundedRect { rect, FloatRoundedRect::Radii { cornerRadius, cornerRadius } });
+    return { path, rect, cornerRadius, CornerType::Noncontinuous };
```
Maybe `WTFMove(path)`

aprotyas replied:
(doesn't fall for the RVO antipattern because it's calling the `RoundedShape` ctor first)

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+    controlRadius = 6.f * zoomScale;
+#else
+    static constexpr auto largeButtonHeight = 45;
+    static constexpr auto largeButtonRadiusRatio = 0.35f / 2;
```
nit: Our style guide suggests dropping the `f` prefix unless required https://webkit.org/code-style-guidelines/#float-suffixes
```suggestion
    static constexpr auto largeButtonRadiusRatio = 0.35 / 2;
```

aprotyas replied:
Same thing with a few other literals in this method.

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+        // If trying to make the button pill-shaped would make it a circle
+        // or nearly circle, use the non-pill shape instead.
+        const auto sizeRatio = rect.width() / rect.height();
+        if (3 / 2.f > sizeRatio && sizeRatio > 2 / 3.f)
```
Might be worth pulling out 1.5 and 0.67 into named constants? I'm not sure what they mean tbh, maybe it's not that important though.

lilyspiniolas replied:
As the width/height ratio approaches 1, a corner radius of `std::min(width, height) / 2.f` makes the control shape approach a circle. Having buttons with equal or similar width/height values turn into circles or near-circles when the form control refresh is enabled seems undesirable, so the pill-shape will only apply if one of the lengths is at least 1.5x larger than the other. IMO having the ratios displayed makes this more clear

pxlcoder replied:
In that case, maybe 1.5 could be the constant, and you could use 1/x in the other?

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+constexpr auto searchFieldDecorationEmSize = 1.f;
+#if PLATFORM(MAC)
+constexpr auto searchFieldDecorationWithDropdownEmSizeLTR = 1.5f;
+constexpr auto searchFieldDecorationWithDropdownEmSizeRTL = 1.7f;
```
`static constexpr`?

aprotyas replied:
also a few more places where we can drop the `f` suffix

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+    const auto isInlineFlipped = style->writingMode().isInlineFlipped();
+    const auto sizeWithDropdown = isInlineFlipped ? searchFieldDecorationWithDropdownEmSizeRTL : searchFieldDecorationWithDropdownEmSizeLTR;
+
+    const auto effectiveSearchFieldDecorationEmSize = supportsResults ? sizeWithDropdown : earchFieldDecorationEmSize;
```
minor typo
```suggestion
    const auto effectiveSearchFieldDecorationEmSize = supportsResults ? sizeWithDropdown : searchFieldDecorationEmSize;
```

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+    const auto effectiveSearchFieldDecorationEmSize = searchFieldDecorationEmSize;
+#endif
+
+    constexpr auto searchFieldInlinePaddingEmSize = 0.55f;
```
```suggestion
    static constexpr auto searchFieldInlinePaddingEmSize = 0.55f;
```

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
+    if (textGapEmSize * pixelsPerEm < borderRadius)
+        return false;
+
+    return true;
```
```suggestion
    return borderRadius <= textGapEmSize * pixelsPerEm;
```

aprotyas reviewed (commented):
This is great! I left some comments inline. I just noticed I commented on some existing code you were just moving around, sorry about that.

pxlcoder commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -632,14 +639,72 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
+                    if (!clipOffset.isZero())
+                        adjustedPath.translate(clipOffset);
+
+                    auto intersectingPath = adoptCF(CGPathCreateCopyByIntersectingPath(adjustedPath.platformPath(), clipPath->platformPath(), false));
```
Explicit `RetainPtr`.

pxlcoder commented on Source/WebCore/page/InteractionRegion.h:
```diff
@@ -62,6 +62,9 @@ struct InteractionRegion {
     std::optional<Path> clipPath { std::nullopt };
+#if ENABLE(FORM_CONTROL_REFRESH)
+    bool useContinuousCorners { false };
+#endif
```
I don't think this needs to be tied to `ENABLE(FORM_CONTROL_REFRESH)`.

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -328,6 +328,13 @@ void EventRegionContext::shrinkWrapInteractionRegions()
+                    // included in the final clip.
+                    Path path;
+                    path.addContinuousRoundedRect(discoveredRegion.rectInLayerCoordinates, discoveredRegion.cornerRadius);
+                    discoveredClipPaths.append(path);
```
👍

lilyspiniolas committed:
[Form Control Refresh][visionOS] Improve gaze glow for updated native control shapes
https://bugs.webkit.org/show_bug.cgi?id=295388
rdar://154134813

Reviewed by Aditya Keerthi.

In RenderThemeCocoa, controls which should have adjusted interaction
regions now each have a method which returns the shape of the control.
These methods are used both during painting, and during the computation
of interaction regions.

In `interactionRegionForRenderedRegion`, we now check if the region
belongs to a control with native appearance. If it does, we
call new method `RenderThemeCocoa::shapeForInteractionRegion`, which
provides information about the control's shape that we can use in order
to adjust the interaction region. This includes the size, corner radius,
path (if needed, only required if a clip path is present), and whether or
not the region should be rendered using continuous corners or typical
rounded corners. When a clip path is already present, we update it to be
the intersecting path between the existing clip path and the bounding path
for the control.

Added layout tests LayoutTests/interaction-region/form-control-refresh/*
which test the adjusted region shapes for native controls by dumping the
regions for a native control, a zoomed native control, a clipped native
control, and a native control that is both zoomed and clipped.

* LayoutTests/interaction-region/button-in-link-expected.txt:
* LayoutTests/interaction-region/clip-path-with-label-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/form-control-refresh/button-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/button-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/checkbox-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/checkbox-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/date-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/date-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/radio-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/radio-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/search-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/search-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/select-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/select-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/slider-thumb-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/slider-thumb-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/switch-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/switch-native-interaction-region.html: Added.
* LayoutTests/interaction-region/form-control-refresh/textfield-native-interaction-region-expected.txt: Added.
* LayoutTests/interaction-region/form-control-refresh/textfield-native-interaction-region.html: Added.
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/labels-expected.txt:
* LayoutTests/interaction-region/select-element-expected.txt:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/platform/visionos-2/TestExpectations:
* LayoutTests/platform/visionos-2/interaction-region/button-in-link-expected.txt: Copied from LayoutTests/interaction-region/button-in-link-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/clip-path-with-label-expected.txt: Copied from LayoutTests/interaction-region/clip-path-with-label-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/event-region-overflow-expected.txt: Copied from LayoutTests/interaction-region/event-region-overflow-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/icon-inside-button-single-region-expected.txt: Copied from LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/input-type-file-region-expected.txt: Copied from LayoutTests/interaction-region/input-type-file-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/input-type-range-region-expected.txt: Copied from LayoutTests/interaction-region/input-type-range-region-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/labels-expected.txt: Copied from LayoutTests/interaction-region/labels-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/select-element-expected.txt: Copied from LayoutTests/interaction-region/select-element-expected.txt.
* LayoutTests/platform/visionos-2/interaction-region/text-input-expected.txt: Copied from LayoutTests/interaction-region/text-input-expected.txt.
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::drawRect):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
(WebCore::operator<<):
* Source/WebCore/page/InteractionRegion.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::addContinuousRoundedRect):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustedPaintRect const):
(WebCore::paintCheckboxRadioInnerShadowForVectorBasedControls):
(WebCore::continuousRoundedShape):
(WebCore::roundedShape):
(WebCore::nodeIsDateOrTimeRelatedInput):
(WebCore::shapeForButton):
(WebCore::shapeForCheckbox):
(WebCore::shapeForRadioButton):
(WebCore::searchFieldCanBeCapsule):
(WebCore::shapeForSearchField):
(WebCore::shapeForSwitchTrack):
(WebCore::shapeForSliderThumb):
(WebCore::shapeForTextAreaOrTextField):
(WebCore::RenderThemeCocoa::paintCheckboxForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintRadioForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSliderThumbForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSearchFieldForVectorBasedControls):
(WebCore::RenderThemeCocoa::shapeForInteractionRegion):
(WebCore::RenderThemeCocoa::inflateRectForInteractionRegion):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustedPaintRect const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):

Canonical link: https://commits.webkit.org/297559@main
